File: python/array_hashmaps/find_the_prefix_common_array_of_two_arrays_leetcode_2656.py
from collections import defaultdict
from typing import List
class Solution:
    def findThePrefixCommonArray(self, A: List[int], B: List[int]) -> List[int]:
        counter = 0
        ans = []
        my_map = defaultdict()
        for i in range(len(A)):
            my_map[A[i]] = my_map.get(A[i], 0) + 1
            my_map[B[i]] = my_map.get(B[i], 0) + 1
            for val in my_map.values():
                counter += val // 2
            ans.append(counter)
            counter = 0
        return ans

if __name__ == "__main__":
    a = [1,3,2,4]
    b = [3,1,2,4]
    sol = Solution()
    c = sol.findThePrefixCommonArray(a, b)
    print(c)

File: cpp/array_hashmaps/find_the_prefix_common_array_of_two_arrays_leetcode_2656.cpp
#include <iostream>
#include <bits/stdc++.h>
using namespace std;
class Solution {
public:
    vector<int> findThePrefixCommonArray(vector<int>& A, vector<int>& B) {
        int counter = 0;
        vector<int> ans;
        unordered_map<int, int> my_map;
        for(int i = 0; i < A.size(); i++){
          my_map[A[i]]++;
          my_map[B[i]]++;
          for(auto iter: my_map){
            counter += iter.second/2;
          }
          ans.push_back(counter);
          counter = 0;
        }
        return ans;
    }
};

int main(){
  vector<int> a = {1,3,2,4};
  vector<int> b = {3,1,2,4};
  Solution sol;
  vector<int> c = sol.findThePrefixCommonArray(a, b);
  for(int i : c){
    std::cout << i << " ";
  }
}
